### DISCUSSION TITLE
A way to speed up vcpkg download. 加速 vcpkg 下载

### DISCUSSION THREAD

**Discussion by cathaysia** [microsoft/vcpkg#17535] [Show and tell]
Chinese | [English](https://github.com/microsoft/vcpkg/discussions/17535#discussioncomment-699204)

English translation thanks to @pratikpc

---

原理很简单，就是本地开个代理，然后将 vcpkg 请求的 host 和 url 改成加速网站的，目前我这边 ghproxy.com 速度最佳，下面我用 [Fiddler](https://www.telerik.com/fiddler) 做代理，用 nginx 等应该也行，但是这些我就不是很清楚了。

打开 Fiddler，然后点击 《工具->选项->HTTPS》，将以下内容打勾：
![15920147-1ba05514fd6bb5b8](https://user-images.githubusercontent.com/42114817/116232268-7d0a0200-a78c-11eb-90e7-6e2aa1b7099b.png)

如果这期间提示要安装 ssl 证书，同意即可。

然后将 Fiddler 的端口改成 8888（任意，我这里默认就是 8888）
![15920147-688b0ecc325da810](https://user-images.githubusercontent.com/42114817/116232391-a6c32900-a78c-11eb-940a-a4dc320b731a.png)

然后点击菜单栏的 规则 -> 自定义规则。 找打 `onBeforeRequest` 函数，添加下面两行函数：

```c#
if(oSession.host == "github.com:443" || oSession.host == "raw.githubusercontent.com:443"){
    oSession.host = "ghproxy.com:443"
}
if(oSession.host == "github.com" || oSession.host == "raw.githubusercontent.com"){
    //oSession.host = "ghproxy.com"
    oSession.fullUrl = "https://ghproxy.com/https://" + oSession.url
}

```

然后在 CMD 中执行 
```cmd
set HTTPS_PROXY=localhost:8888
set HTTP_PROXY=localhost:8888
vcpkg install ...
```
shell 同理，只要设置两个环境变量就行了

> 注意：
>
> 1. 等号两边不要加空格
> 2. 在 5 月 5 号的版本, vcpkg 已经可以自动检测代理了。 感谢 @cnSchwarzer
>     参阅： https://github.com/microsoft/vcpkg/discussions/17535#discussioncomment-709949

。。。
需要注意的是：由于 Fiddler 会自动拦截浏览器的请求，因此在 vcpkg 安装的过程中可能会导致浏览器无法访问 github，这是由于 Fiddler 拦截请求导致的。安装完后关闭 Fiddler 即可。

**Comment by pratikpc** [comment]
@cathaysia I am not a Chinese speaker but was able to use Bing Translate to get this translated  
This is [fiddler BTW](https://en.wikipedia.org/wiki/Fiddler_(software))
Translated text below

---

emm, I'm not good at English, expect a translation.
Chinese below.

The principle is very simple, is to open a local agent, and then vcpkg request host and url to speed up the site, my side ghproxy.com the best speed, below I use Fiddler as an agent, with nginx and so on should be OK, but these I am not very clear.

Open Fiddler, and then click Tools - Options - HTTPS to tick the following:
15920147-1ba05514fd6bb5b8

If you are prompted to install the ssl certificate during this period, agree.

Then change the port of Fiddler to 8888 (any, I'll default here to 8888)
15920147-688b0ecc325da810

Then click on the rules in the menu bar - and customize the rules. Look for the function and add the following two lines of function:onBeforeRequest

if(oSession.host == "github.com:443" || oSession.host == "raw.githubusercontent.com:443"){
    oSession.host = "ghproxy.com:443"
}
if(oSession.host == "github.com" || oSession.host == "raw.githubusercontent.com"){
    //oSession.host = "ghproxy.com"
    oSession.fullUrl = "https://ghproxy.com/https://" + oSession.url
}
It is then executed in the CMD

set HTTPS_PROXY = localhost:8888
set HTTP_PROXY  = localhost:8888
vcpkg install ...
Shell, by the same to name, just set two environment variables

。。。
Note: Because Fiddler automatically blocks browser requests, it may cause the browser to lose access to the github during vcpkg installation, which is due to Fiddler blocking requests. Turn off Fiddler after installation.

I put this article https://www.jianshu.com/p/4dfbe353be0f. But the article is locked and I don't know if it was unsealed by the time you looked it up

**Reply by cathaysia** [reply]
Thanks very much! Here some meaning not covered, it seem just because the terms are not same between Chinese and English.

----

The principle is very simple, is to open a local agent to modify all the host and url that  vcpkg used to the mirror-site, ghproxy.com is the best for me. Other mirror site should be OK. Below I use [Fiddler](https://www.telerik.com/fiddler) as an agent, with nginx and so on should be OK, but these I am not very clear.

Open Fiddler, and then click Tools - Options - HTTPS to tick the following:

![15920147-1ba05514fd6bb5b8](https://user-images.githubusercontent.com/42114817/116232268-7d0a0200-a78c-11eb-90e7-6e2aa1b7099b.png)

If you are prompted to install the ssl certificate during this period, agree.

Then change the port of Fiddler to 8888 (any, the default port is 8888)

![15920147-688b0ecc325da810](https://user-images.githubusercontent.com/42114817/116232391-a6c32900-a78c-11eb-940a-a4dc320b731a.png)

Then click on the rules in the menu bar - and customize the rules. Find the function `onBeforeRequest ` and add the following code:

```c#
if(oSession.host == "github.com:443" || oSession.host == "raw.githubusercontent.com:443"){
oSession.host = "ghproxy.com:443"
}
if(oSession.host == "github.com" || oSession.host == "raw.githubusercontent.com"){
//oSession.host = "ghproxy.com"
oSession.fullUrl = "https://ghproxy.com/https://" + oSession.url
}
```

For CMD (on Windows)

```cmd
set HTTPS_PROXY=localhost:8888
set HTTP_PROXY=localhost:8888
vcpkg install ...
```

Shell, in the similar way, just set two environment variables `HTTPS_PROXY ` and `HTTP_PROXY `

> Note:
>
> 1. no space surround "="
> 2. After the version 2021-05-05 version, vcpkg has the ability to auto use proxy. Thanks to @cnSchwarzer 
>  see: https://github.com/microsoft/vcpkg/discussions/17535#discussioncomment-709949

...
Note: Because Fiddler automatically blocks browser requests, it may cause the browser to lose access to the github during Fiddler run, which is due to Fiddler blocking requests. Turn off Fiddler should be OK.

---

**Comment by reitowo** [comment]
更新到5月5号的版本，可以使用自动代理检测了，详见PR：https://github.com/microsoft/vcpkg-tool/pull/49

English Translation: Please update to vcpkg.exe version 2021-05-05 (by `git pull` and `./bootstrap-vcpkg.bat`), I added an IE proxy auto integration feature, see https://github.com/microsoft/vcpkg-tool/pull/49

All you need to do is ensure this Windows 10 proxy setting page is set by your proxy software like this:
![image](https://user-images.githubusercontent.com/29846655/117529773-20200e80-b00c-11eb-80a2-0669a4c22667.png)

It will then automatically detects it (called IE Proxy Setting) and use it for download.
![image](https://user-images.githubusercontent.com/29846655/117529810-69705e00-b00c-11eb-870a-7ebf2f7ccf4f.png)

**Comment by JackBoosY** [comment]
其实mirror的PR好了我们就可以从国内直接下载了。

**Reply by reitowo** [reply]
那么多仓库哪来的mirror

**Reply by JackBoosY** [reply]
会有命令仅下载所有工具和packages的，只需要上传国内服务器就行。
以前我弄了一个，不过下载太麻烦，更新又很快所以只维持了一年。

**Reply by cathaysia** [reply]
Github 在国内主要是 host 混淆 + 443 端口屏蔽。 如果自己手动改了 hosts 之后，使用 wget 可以在一定程度上从 github 上下载文件（前提是 443 端口能正常访问）。但是使用第三方工具 vcpkg 貌似还不支持。

如果将 wget 的证书校验关掉的话，可以简单地将对 github.com:443 的访问重定向到任意一个网站的 443 端口。但是 wget 基本上不走代理不知道什么情况

github.com 的 host 可以去 ipaddress.com 上自己查，或者是用 https://github.com/cathaysia/anslookup/releases 这个软件。

**Comment by cathaysia** [comment]
@cnSchwarzer 你确定那个 PR 可以正常工作吗？我使用新版本的时候发现 Fiddler 一直捕获不到 vcpkg 的流量。使用二月份版本的时候倒是能正常工作。

可否只在用户没有设置 HTTP_PROXY 的时候设置它？不要让它覆盖我手动设置的变量？

**Reply by reitowo** [reply]
检查你的Windows 10代理设置页，如果上图里的使用代理服务器没有开，那我的PR就不会检测

**Reply by reitowo** [reply]
但正如你所说我忘记检测环境变量是否已经有HTTP_PROXY了，我再提交个PR改下。。

**Comment by JackBoosY** [comment]
For Chinese users:
对于中国用户，镜像在 https://gitee.com/jackboosy/vcpkg/issues/I44496
